File: Rust/src/1.two-sum.rs
/*
 * @lc app=leetcode id=1 lang=rust
 *
 * [1] Two Sum
 */

// @lc code=start
use std::collections::HashMap;
//impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut hm=HashMap::new();
        let mut ans: Vec<i32>=Vec::new();
        for (i,el) in nums.iter().enumerate() {
            //println!("{} {}",i,el);
            let search=target-el;
            if hm.contains_key(&search) {
                ans.push(i as i32);
                ans.push(hm[&search]);
                ans.sort();
                return ans;
            }
            hm.insert(el,i as i32);
        }
        ans

    }
//}
// @lc code=end

#[cfg(test)]
mod tests {
    use super::two_sum;

    #[test]
    fn test_two_sum() {
        // Standard case.
        assert_eq!(two_sum(vec![1, 2, 3, 4, 5], 9), vec![3, 4]);
    }
}

File: Python/1.TwoSum.py
class Solution:
    def twoSum(self, nums, target: int):
        dct={} #num: idx
        for idx,n in enumerate(nums):
            searched=target-n
            if searched in dct:
                return [dct[searched], idx]
            dct[n]=idx

s=Solution()
print(s.twoSum([2,7,11,15],9)) # [0,1] (2+7=9)
print(s.twoSum([3,2,4],6)) # [1,2] 
print(s.twoSum([3,3],6)) # [0,1] 